go/webpack/signutil.go

```diff
@@ -0,0 +1,160 @@
+package webpack
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rand"
+	"crypto/rsa"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"math/big"
+)
+
+// Sign() signs message using a TLS1.3 algorithm chosen depending on pk:
+//   RSA <= 3072 bits:    rsa_pss_sha256
+//   RSA, 3073-7680 bits: rsa_pss_sha384
+//   secp256r1:           ecdsa_secp256r1_sha256
+//   secp384r1:           ecdsa_secp384r1_sha384
+//   secp521r1:           ecdsa_secp521r1_sha512
+func Sign(pk crypto.PrivateKey, message []byte) ([]byte, error) {
+	signer, err := signerForPrivateKey(pk)
+	if err != nil {
+		return nil, err
+	}
+	return signer.sign(message)
+}
+
+// Verify() verifies that signature signs message using pk, using the algorithm
+// identified by Sign(). Returns nil for when signature is valid.
+func Verify(pk crypto.PublicKey, message []byte, signature []byte) error {
+	verifier, err := verifierForPublicKey(pk)
+	if err != nil {
+		return err
+	}
+	return verifier.verify(message, signature)
+}
+
+type messageSigner interface {
+	sign(message []byte) ([]byte, error)
+}
+
+func signerForPrivateKey(pk crypto.PrivateKey) (messageSigner, error) {
+	switch pk := pk.(type) {
+	case *rsa.PrivateKey:
+		switch bits := pk.N.BitLen(); {
+		case bits <= 3072:
+			return rsaPSSSigner{pk, crypto.SHA256}, nil
+		case bits <= 7680:
+			return rsaPSSSigner{pk, crypto.SHA384}, nil
+		default:
+			return nil, fmt.Errorf("RSA key too big: %v bits", bits)
+		}
+	case *ecdsa.PrivateKey:
+		switch name := pk.Curve.Params().Name; name {
+		case "P-256":
```

Comment:
I wish these were pre-defined constants somewhere. But I didn't see that in a brief overview of crypto/elliptic